SmartDeviceLink/SDLTouchManager.m

```diff
@@ -0,0 +1,241 @@
+//  Copyright © 2016 smartdevicelink. All rights reserved.
+//
+
+#import "SDLTouchManager.h"
+
+#import "dispatch_timer.h"
+#import "SDLOnTouchEvent.h"
+#import "SDLPinchGesture.h"
+#import "SDLProxyListener.h"
+#import "SDLTouch.h"
+#import "SDLTouchCoord.h"
+#import "SDLTouchEvent.h"
+#import "SDLTouchType.h"
+
+typedef NS_ENUM(NSUInteger, SDLPerformingTouchType) {
+    SDLPerformingTouchTypeNone,
+    SDLPerformingTouchTypeSingleTouch,
+    SDLPerformingTouchTypeMultiTouch,
+    SDLPerformingTouchTypePanningTouch
+};
+
+static NSUInteger const kMaximumNumberOfTouches = 2;
+
+@interface SDLTouchManager () <SDLProxyListener>
+
+@property SDLTouch previousTouch;
+
+/*
+ * Only used for caching previous single taps for double tap detection
+ */
+@property SDLTouch singleTapTouch;
+
+@property CGFloat previousPinchDistance;
+
+@property SDLPinchGesture currentPinchGesture;
+
+@property dispatch_source_t singleTapTimer;
+
+@property SDLPerformingTouchType performingTouchType;
+
+@end
+
+@implementation SDLTouchManager
+
+- (instancetype)init {
+    if (self = [super init]) {
+        _panTimeThreshold = 50.0f;
+        _tapTimeThreshold = 0.4f;
+        _tapDistanceThreshold = 50.0f;
+        _previousTouch = SDLTouchZero;
+        _singleTapTouch = SDLTouchZero;
+        _touchEnabled = YES;
+    }
+    return self;
+}
+
+#pragma mark - SDLProxyListener Delegate
+- (void)onProxyOpened { }
+- (void)onProxyClosed { }
+- (void)onOnHMIStatus:(SDLOnHMIStatus *)notification { }
+- (void)onOnDriverDistraction:(SDLOnDriverDistraction *)notification { }
+
+- (void)onOnTouchEvent:(SDLOnTouchEvent *)notification {
+    if (!self.isTouchEnabled) {
+        return;
+    }
+    
+    SDLTouchEvent* touchEvent = notification.event.firstObject;
+    NSUInteger touchEventID = touchEvent.touchEventId.unsignedIntegerValue;
+    
+    if (touchEventID > kMaximumNumberOfTouches) {
+        return;
+    }
+    
+    SDLTouchCoord* coord = touchEvent.coord.firstObject;
+    
+    NSNumber* timeStampNumber = touchEvent.timeStamp.firstObject;
+    NSUInteger timeStamp;
+    // In the event we receive a null timestamp, we will supply a device timestamp.
+    if (timeStampNumber != [NSNull null]) {
+        timeStamp = timeStampNumber.unsignedIntegerValue;
+    } else {
+        timeStamp = [[NSDate date] timeIntervalSince1970] * 1000.0;
+    }
+    
+    SDLTouch touch = SDLTouchMake(touchEventID,
+                                  coord.x.floatValue,
+                                  coord.y.floatValue,
+                                  timeStamp);
+    if ([notification.type isEqualToEnum:SDLTouchType.BEGIN]) {
+        [self sdl_handleBeginTouch:touch];
+    } else if ([notification.type isEqualToEnum:SDLTouchType.MOVE]) {
+        [self sdl_handleMovedTouch:touch];
+    } else if ([notification.type isEqualToEnum:SDLTouchType.END]) {
+        [self sdl_handleEndTouch:touch];
+    }
+}
+
+#pragma mark - Private
+- (void)sdl_handleBeginTouch:(SDLTouch)touch {
```

Comment:
Naming should probably be `sdl_handleTouchBegin` since that helps make the other, similar method names appear more similar to the eye even though it sounds kinda weird audibly.